src/analyses/does_remove_const.cpp

```diff
@@ -0,0 +1,138 @@
+/*******************************************************************\
+
+ Module: Analyses
+
+ Author: DiffBlue Limited. All rights reserved.
+
+\*******************************************************************/
+
+#include <goto-programs/goto_program.h>
+#include <util/type.h>
+#include <util/expr.h>
+#include <util/std_code.h>
+#include <util/base_type.h>
+#include <ansi-c/c_qualifiers.h>
+
+#include "does_remove_const.h"
+
+/*******************************************************************\
+
+Function: does_remove_constt::does_remove_constt
+
+  Inputs:
+   goto_program - the goto program to check
+   ns - the namespace of the goto program (used for checking type equality)
+
+ Outputs:
+
+ Purpose: A naive analysis to look for casts that remove const-ness from
+          pointers.
+
+\*******************************************************************/
+
+does_remove_constt::does_remove_constt(
+  const goto_programt &goto_program,
+  const namespacet &ns):
+    goto_program(goto_program),
+    ns(ns)
+{}
+
+/*******************************************************************\
+
+Function: does_remove_constt::operator()
+
+  Inputs:
+
+ Outputs: Returns true if the pgroam contains a const-removing cast
+
+ Purpose: A naive analysis to look for casts that remove const-ness from
+          pointers.
+
+\*******************************************************************/
+
+bool does_remove_constt::operator()() const
+{
+  for(const goto_programt::instructiont &instruction :
+    goto_program.instructions)
+  {
+    if(instruction.is_assign())
+    {
+      const code_assignt assign=to_code_assign(instruction.code);
+      typet rhs_type=assign.rhs().type();
+      typet lhs_type=assign.lhs().type();
+
+      // Compare the types recursively for a point where the rhs is more
+      // const that the lhs
+      if(!is_type_at_least_as_const_as(lhs_type, rhs_type))
+      {
+        return true;
+      }
+
+      // see if the rhs loses const inside the expression tree
+      exprt rhs=assign.rhs();
+      while(!rhs.operands().empty())
+      {
+        typet &pre_op_type=rhs.type();
+        typet &post_op_type=rhs.op0().type();
+
+        // Types equality does not check, for example, const-ness
+        // If this is true, then this expression only modified the
+        // type by some qualifier (or not at all)
+        if(base_type_eq(pre_op_type, post_op_type, ns))
+        {
+          if(!is_type_at_least_as_const_as(pre_op_type, post_op_type))
+          {
+            return true;
+          }
+        }
+        rhs=rhs.op0();
```

Comment:
Yeah so this assignment breaks by only descending `ID_typecast`:

```C
  // Warning: this loses const-ness of f2
  void_fp * p2fp = ((void_fp*)&fp) + 0;
```

And indeed if I swap it around, the original code doesn't even work since it is actually `op1` we need to descend. 

I think we need to check all operands with the same fundamental type as the root, does that feel like a robust approach?